### PR TITLE
DEVX-1292: Producer/consumer example: Spring Boot

### PR DESCRIPTION
**Pull request by gAmUssA**
- [x] java consumer / producer
- [x] confluent local config
- [x] add logo to README
- [x] add README how to test and run (local and cloud)
- [x] cloud config
- [x] Kafka Streams Example

### CODE REVIEW

**Comment by ybyzek** [comment]
🥇 @gAmUssA !  Can you also please add an entry to https://github.com/confluentinc/examples/blob/5.4.0-post/clients/cloud/README.md ?

**Comment by rspurgeon** [comment]
Thanks @gAmUssA , taking a look.  Should we add a README to the directory, similar to the other language examples?

File: clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java
Comment (rspurgeon) [inline clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java]: Only for your consideration @gAmUssA, I thought handling the null case might be nice, but just a suggestion.

```
log.info(
              Optional.of(result)
                .map(r -> r.getRecordMetadata())
                .map(m -> String.format("Produced record to topic {} partition {} @ offset {}",
                  m.topic(), m.partition(), m.offset()))
                .orElse("produce.send result was null"));
```

Comment (gAmUssA) [reply clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java]: I was following java example.

**Comment by gAmUssA** [comment]
> Thanks @gAmUssA , taking a look. Should we add a README to the directory, similar to the other language examples?

File: clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java
Comment (rspurgeon) [reply clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java]: No worries... I'm good w/ this but I'll wait on the README before approval.

File: clients/cloud/java-springboot/src/main/resources/log4j.properties
Comment (bbejeck) [inline clients/cloud/java-springboot/src/main/resources/log4j.properties:20]: Just out of curiosity why `System.err` vs `System.out`, I'm fine with this as-is, just asking out of curiosity.

**Review by bbejeck — COMMENTED** [review]
LGTM, modulo the comments from @rspurgeon.

File: clients/cloud/java-springboot/src/main/resources/log4j.properties
Comment (edenhill) [reply clients/cloud/java-springboot/src/main/resources/log4j.properties:20]: I think this is good, logs should not interfere with the program's output.

File: clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java
Comment (gAmUssA) [reply clients/cloud/java-springboot/src/main/java/io/confluent/examples/clients/cloud/springboot/ProducerExample.java]: @rspurgeon working on it

**Comment by gAmUssA** [comment]
@bbejeck @rspurgeon Gents, kindly review. It's ready for merge

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: @gAmUssA all the other examples use a standard Java properties file with just the key properties to connect to a cluster.  

```
$ cat $HOME/.ccloud/config
bootstrap.servers=<BROKER ENDPOINT>
ssl.endpoint.identification.algorithm=https
security.protocol=SASL_SSL
sasl.mechanism=PLAIN
sasl.jaas.config=org.apache.kafka.common.security.plain.PlainLoginModule required username\="<API KEY>" password\="<API SECRET>";
```

However, this PR deviates from that.  Can this example follow the other examples?  If Spring needs `.application-ccloud.properties`, then the example should auto-generate it from the standard Java properties file.

Reasons to do it this way:

1. Consistency
2. Eventually Confluent CLI will autogenerate the config file, so this example can just consume it.

File: clients/cloud/README.md
Comment (ybyzek) [inline clients/cloud/README.md:16]: Please change `springboot.png` to be an icon of just Spring (remove SR and Avro).

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: There should be two separate examples: 

1. One without Avro/SR
2. One with Avro/SR

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: @gAmUssA for consistency, please use the same formatting as described in another example, e.g. the Java one

**Review by ybyzek — CHANGES_REQUESTED** [review]
@gAmUssA thanks for the PR.  Left feedback on some key things that should be addressed before merge, thanks!

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: I would submit separate PR for JSON version of this example.

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: same formatting for configuration file or README?

Comment (ybyzek) [reply clients/cloud/java-springboot/README.adoc]: Should be both: -- same configuration file can be used across all examples, and README so that the text/example is the same

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: it's done!

File: clients/cloud/README.md
Comment (gAmUssA) [reply clients/cloud/README.md:16]: done

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc:6]: @gAmUssA unfortunately the section Prerequisites format has changed since this PR was opened.  Can you please make it similar to https://github.com/confluentinc/examples/blob/5.4.1-post/clients/cloud/groovy/README.adoc ?

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: @gAmUssA all client examples use a common config file as describe in https://github.com/confluentinc/configuration-templates/blob/master/README.md

Which means that this `application-ccloud.properties` file must be auto-generated from that common config file.  The value in doing it this way:

- provides consistency between examples
- enables users to try different code languages using the same config file
- enables us to test

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: `gradle` is a prereq

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: `<1>` (and numbers below) -- what is their purpose?  It doesn't work with them so they should be removed

**Review by ybyzek — CHANGES_REQUESTED** [review]

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: we use Gradle wrapper. Gradle isn't prerequisite in this case, because wrapper will download required version of gradle

File: clients/cloud/java-springboot/scripts/ccloud-generate-cp-configs.sh
Comment (ybyzek) [inline clients/cloud/java-springboot/scripts/ccloud-generate-cp-configs.sh]: This file is already in this repo at https://github.com/confluentinc/examples/blob/5.4.1-post/ccloud/ccloud-generate-cp-configs.sh, can you just use the existing one?

Comment (gAmUssA) [reply clients/cloud/java-springboot/scripts/ccloud-generate-cp-configs.sh]: done. switched to use standard scripts

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc:6]: updated

**Comment by gAmUssA** [comment]
@rspurgeon why did you close it, Rick?

**Comment by rspurgeon** [comment]
@gAmUssA I think me closing this was a 'fat finger'... my bad

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
NOTE: Produce and consume messages from [Confluent Cloud](https://www.confluent.io/confluent-cloud/?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud) using the Spring Boot Java Framework.
```

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
* Java 1.8
```

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
NOTE: To build and execute all examples run `start.sh` and use `stop.sh` to cleanup.
```

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc:29]:
```suggestion
This example uses values that are formatted in Avro and integrates with the Confluent Cloud Schema Registry.
```

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
Before using Confluent Cloud Schema Registry, check [availability and limits](https://docs.confluent.io/current/cloud/limits.html?utm_source=github&utm_medium=demo&utm_campaign=ch.examples_type.community_content.clients-ccloud), and be sure you have connectivity to the Confluent Cloud Schema Registry public endpoint.
```

**Comment by rspurgeon** [comment]
On my machine, it doesn't appear that the application runs syncronously to the rest of the script.  Notice how my prompt shows up in between Starting Spring Boot application (Kafka Streams)... and the Spring logo

```

        ☁️  Building Spring Boot application...
Downloading https://services.gradle.org/distributions/gradle-6.0.1-all.zip
......................................................................................................................................
Starting a Gradle Daemon (subsequent builds will be faster)

BUILD SUCCESSFUL in 1m 16s
4 actionable tasks: 4 executed
        ☁️  Starting Spring Boot application (Vanila Kafka API)...
        ☁️  Starting Spring Boot application (Kafka Streams)...

09:27:22 in examples/clients/cloud/java-springboot on  DEVX-1292 [?] took 1m 17s
[I] ➜
  .   ____          _            __ _ _
 /\\ / ___'_ __ _ _(_)_ __  __ _ \ \ \ \
( ( )\___ | '_ | '_| | '_ \/ _` | \ \ \ \
 \\/  ___)| |_)| | | | | || (_| |  ) ) ) )
  '  |____| .__|_| |_|_| |_\__, | / / / /
 =========|_|==============|___/=/_/_/_/
 :: Spring Boot ::        (v2.2.4.RELEASE)


```

**Comment by rspurgeon** [comment]
I get the following exception when running against my CCloud which requires RF of 3:

```
2020-04-09 09:44:01.341 ERROR 62340 --- [           main] o.springframework.kafka.core.KafkaAdmin  : Could not configure topics

org.springframework.kafka.KafkaException: Failed to create topics; nested exception is org.apache.kafka.common.errors.PolicyViolationException: Topic replication factor must be 3
        at org.springframework.kafka.core.KafkaAdmin.addTopics(KafkaAdmin.java:266) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
        at org.springframework.kafka.core.KafkaAdmin.addTopicsIfNeeded(KafkaAdmin.java:206) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
        at org.springframework.kafka.core.KafkaAdmin.initialize(KafkaAdmin.java:174) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
        at org.springframework.kafka.core.KafkaAdmin.afterSingletonsInstantiated(KafkaAdmin.java:144) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
        at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:896) [spring-beans-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:878) [spring-context-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:550) [spring-context-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
        at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
        at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
        at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
        at io.confluent.examples.clients.cloud.springboot.kafka.SpringbootKafkaApplication.main(SpringbootKafkaApplication.java:31) [classes!/:na]
        at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_242]
        at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_242]
        at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_242]
        at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_242]
        at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [java-springboot-0.0.1-SNAPSHOT.jar:na]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [java-springboot-0.0.1-SNAPSHOT.jar:na]
        at org.springframework.boot.loader.Launcher.launch(Launcher.java:51) [java-springboot-0.0.1-SNAPSHOT.jar:na]
        at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:597) [java-springboot-0.0.1-SNAPSHOT.jar:na]
Caused by: org.apache.kafka.common.errors.PolicyViolationException: Topic replication factor must be 3

```

**Comment by gAmUssA** [comment]
@rspurgeon apps are running asynchronously (need to send them to bg to retrieve PID so `stop.sh` script should work)

**Comment by gAmUssA** [comment]
> I get the following exception when running against my CCloud which requires RF of 3:
> 
> ```
> 2020-04-09 09:44:01.341 ERROR 62340 --- [           main] o.springframework.kafka.core.KafkaAdmin  : Could not configure topics
> 
> org.springframework.kafka.KafkaException: Failed to create topics; nested exception is org.apache.kafka.common.errors.PolicyViolationException: Topic replication factor must be 3
>         at org.springframework.kafka.core.KafkaAdmin.addTopics(KafkaAdmin.java:266) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
>         at org.springframework.kafka.core.KafkaAdmin.addTopicsIfNeeded(KafkaAdmin.java:206) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
>         at org.springframework.kafka.core.KafkaAdmin.initialize(KafkaAdmin.java:174) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
>         at org.springframework.kafka.core.KafkaAdmin.afterSingletonsInstantiated(KafkaAdmin.java:144) [spring-kafka-2.3.5.RELEASE.jar!/:2.3.5.RELEASE]
>         at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:896) [spring-beans-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
>         at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:878) [spring-context-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
>         at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:550) [spring-context-5.2.3.RELEASE.jar!/:5.2.3.RELEASE]
>         at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
>         at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
>         at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) [spring-boot-2.2.4.RELEASE.jar!/:2.2.4.RELEASE]
>         at io.confluent.examples.clients.cloud.springboot.kafka.SpringbootKafkaApplication.main(SpringbootKafkaApplication.java:31) [classes!/:na]
>         at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_242]
>         at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_242]
>         at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_242]
>         at java.lang.reflect.Method.invoke(Method.java:498) ~[na:1.8.0_242]
>         at org.springframework.boot.loader.MainMethodRunner.run(MainMethodRunner.java:48) [java-springboot-0.0.1-SNAPSHOT.jar:na]
>         at org.springframework.boot.loader.Launcher.launch(Launcher.java:87) [java-springboot-0.0.1-SNAPSHOT.jar:na]
>         at org.springframework.boot.loader.Launcher.launch(Launcher.java:51) [java-springboot-0.0.1-SNAPSHOT.jar:na]
>         at org.springframework.boot.loader.PropertiesLauncher.main(PropertiesLauncher.java:597) [java-springboot-0.0.1-SNAPSHOT.jar:na]
> Caused by: org.apache.kafka.common.errors.PolicyViolationException: Topic replication factor must be 3
> ```

do you use `./start.sh` script to run it? it looks like app pics up wrong profile

**Comment by gAmUssA** [comment]
@rspurgeon I pushed some stuff that potentially eliminates your issue

**Comment by rspurgeon** [comment]
> @rspurgeon apps are running asynchronously (need to send them to bg to retrieve PID so `stop.sh` script should work)

The issue with this approach is that you return control of the terminal to the user but are still printing to STDOUT which is, potentially, very confusing.   AFAIK, all our other demos that contain applications run synchronously and wait for an expected number of records to produce/consume.  If you want to run it in the background for some reason, I think we need to pipe the output to a file and then display/direct the user to that output but the user will need to know that the demo is running w/ some other messaging.

**Comment by gAmUssA** [comment]
@rspurgeon what would you suggest? I don't have a strong opinion on the flow of the application.

**Comment by rspurgeon** [comment]
> @rspurgeon what would you suggest? I don't have a strong opinion on the flow of the application.

Would it be an easy change to make the programs produce / consume expected numbers of records and have them exit after that point?  Also, might be good to add a timeout to that so they don't run forever.

**Comment by gAmUssA** [comment]
> > @rspurgeon what would you suggest? I don't have a strong opinion on the flow of the application.
> 
> Would it be an easy change to make the programs produce / consume expected numbers of records and have them exit after that point? Also, might be good to add a timeout to that so they don't run forever.

not sure if I like this approach but if you say this make this example merged I will do that

**Comment by rspurgeon** [comment]
> not sure if I like this approach but if you say this make this example merged I will do that

If there is another approach that is fine, I just don't think giving the user control over the terminal while applications print to STDOUT works.

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
Run producer and consumer application by executing `./startProducerConsumer.sh`.
```

File: clients/cloud/java-springboot/README.adoc
Comment (ybyzek) [inline clients/cloud/java-springboot/README.adoc]: @gAmUssA in GH this doesn't render well.  Should it be just `./startStreams.sh`?

Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: done

File: clients/cloud/java-springboot/README.adoc
Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: done

Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]:
```suggestion
Run producer and consumer application by executing:
[source,shell]
----
./startProducerConsumer.sh
----
```

File: clients/cloud/java-springboot/README.adoc
Comment (rspurgeon) [inline clients/cloud/java-springboot/README.adoc]: This shell command looks incorrect to me, can you validate that's what it's supposed to be?

Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: @rspurgeon update your copy

Comment (gAmUssA) [reply clients/cloud/java-springboot/README.adoc]: this command is gone now

**Review by rspurgeon — APPROVED** [review]
LGTM

**Review by ybyzek — APPROVED** [review]
LGTM

**Pull request merged**
